server/src/main/java/io/envoyproxy/controlplane/server/ProtoResourcesSerializer.java

```diff
@@ -0,0 +1,32 @@
+package io.envoyproxy.controlplane.server;
+
+import com.google.common.cache.Cache;
+import com.google.common.cache.CacheBuilder;
+import com.google.protobuf.Any;
+import com.google.protobuf.Message;
+
+import java.util.Collection;
+import java.util.List;
+import java.util.concurrent.ExecutionException;
+import java.util.stream.Collectors;
+
+class ProtoResourcesSerializer {
+
+  private static final Cache<Collection<? extends Message>, List<Any>> cache = CacheBuilder.newBuilder()
+      .weakValues()
```

Comment:
shouldn't this have weakKeys as well?

Reply:
I don't think so.

The behaviour I wanted to achieve is that as long as proto is stored somewhere by someone, keep it in the cache.
`weakKeys()` compares key by the reference. Look at the usage:
`ProtoResourcesSerializer.serialize(response.resources())`
 `response.resources()` does not point to the same list reference across the `send` invocation. The list is built in `SimpleCache#createResponse` and it's different instance for every `Response`.

Reply:
In ADS mode the list is always the full list of resources and is always the same. This is not true for XDS, I think you're right though, having weakValues should be enough